src/java/org/jivesoftware/openfire/pubsub/PubSubEngine.java

```diff
@@ -115,7 +123,14 @@ public boolean process(PubSubService service, IQ iq) {
             action = childElement.element("create");
             if (action != null) {
                 // Entity is requesting to create a new node
-                createNode(service, iq, childElement, action);
+                final Element finalAction = action;
+                // Complete this asynchronously, as UserManager::isRegisteredUser(JID) blocks, waiting for a result which may come in on this thread
+                TaskEngine.getInstance().submit(new Runnable() {
+                    @Override
+                    public void run() {
+                        createNode(service, iq, childElement, finalAction);
+                    }
+                });
                 return true;
```

Comment:
This return statement indicates that the IQ stanza was handled, which, now that things are asynchronous, is no longer the case. I'm not sure how this will hold up in a larger chain of handlers. Even when there are no issues here, I'm not comfortable by setting a precedent here. Perhaps we could make things more explicit, and return a Future-like promise instead?

Reply:
I like this idea; I think (in the absence of returning a Java 8 `Optional<Future>`) we'll have to return `null` (to indicate the packet was not handled) or the `Future` itself where it is.

Reply:
PS When I wrote the original changes, I did check if the existing uses were likely to be thrown off by the async processing, and it all seems to be OK, but making it more explicit is definitely a plus point.